sdks/python/apache_beam/utils/retry.py

```diff
@@ -168,15 +183,29 @@ def retry_on_server_errors_and_timeout_filter(exception):
 def retry_on_server_errors_timeout_or_quota_issues_filter(exception):
   """Retry on server, timeout and 403 errors.
 
-  403 errors can be accessDenied, billingNotEnabled, and also quotaExceeded,
-  rateLimitExceeded."""
+  403 errors include both transient (accessDenied, billingNotEnabled) and
+  non-transient errors (quotaExceeded, rateLimitExceeded). Only retry transient
+  errors."""
   if HttpError is not None and isinstance(exception, HttpError):
     if exception.status_code == 403:
+      try:
+        # attempt to extract the reason and check if it's retryable
+        return exception.content["error"]["errors"][0][
+            "reason"] in _RETRYABLE_REASONS
+      except Exception:
+        _LOGGER.debug(
+            "Could not determine if HttpError is non-transient. Will retry: %s",
+            exception)
       return True
```

Comment:
we shouldn't blindly retry on all `403` exceptions. Forbidden exceptions belongs to permanent exceptions. 

Here's a guide from GCP about healthy retry policies:

https://cloud.google.com/storage/docs/retry-strategy

Reply:
Thanks for this link, I added some of these status codes as retryable.

Looks like another odd behavior is BigQuery can return [`notImplemented` as 501](https://cloud.google.com/bigquery/docs/error-messages#:~:text=notImplemented,for%20more%20information.) 

> we shouldn't blindly retry on all 403 exceptions. Forbidden exceptions belongs to permanent exceptions.

SGTM